# Project 01: Predictive Analytics for Customer Churn
---

Dataset Source:

Moro, S., Rita, P., & Cortez, P. (2014). Bank Marketing [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5K306.
Dataset Description:

This dataset contains customer and campaign information from direct phone marketing campaigns conducted between May 2008 and November 2010. The classification goal is to predict whether a contacted client will subscribe to a term deposit (y). The `bank-additional-full.csv` contains 41,188 rows and 20 input features, very close to the data analyzed in [Moro et al., 2014].

## Load
---

In [1]:
#Libraries
import pandas as pd

In [2]:
#Load dataset
df = pd.read_csv("bank-additional-full.csv", sep=";")

#Preview
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

#### Quick Note: 
The bank-additional-full.csv dataset uses semicolons instead of commas to seperate values. So, to make sure that the column was loaded correctly and the data types read properly, I imported the file using `sep=';'`.

## Exploratory Data Analysis (EDA)
---

In [3]:
#Shape
df.shape

(41188, 21)

In [4]:
#Data Types and basic outline
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [5]:
#Summary Stats
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

In [6]:
#Check Target's Distribution
df['y'].value_counts()

y
no     36548
yes     4640
Name: count, dtype: int64

In [7]:
#Review Categorical Features
categorical_cols = [c for c in df.columns if df[c].dtype == 'object']

for col in categorical_cols:
    print(df[col].value_counts().to_frame())
    print("-" * 10)

               count
job                 
admin.         10422
blue-collar     9254
technician      6743
services        3969
management      2924
retired         1720
entrepreneur    1456
self-employed   1421
housemaid       1060
unemployed      1014
student          875
unknown          330
----------
          count
marital        
married   24928
single    11568
divorced   4612
unknown      80
----------
                     count
education                 
university.degree    12168
high.school           9515
basic.9y              6045
professional.course   5243
basic.4y              4176
basic.6y              2292
unknown               1731
illiterate              18
----------
         count
default       
no       32588
unknown   8597
yes          3
----------
         count
housing       
yes      21576
no       18622
unknown    990
----------
         count
loan          
no       33950
yes       6248
unknown    990
----------
           count
contact         
cellular   2614

In [8]:
#Missing Values
df.isna().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

In [9]:
#Verify unique values and counts `for pdays`
df['pdays'].value_counts().sort_index()

pdays
0         15
1         26
2         61
3        439
4        118
5         46
6        412
7         60
8         18
9         64
10        52
11        28
12        58
13        36
14        20
15        24
16        11
17         8
18         7
19         3
20         1
21         2
22         3
25         1
26         1
27         1
999    39673
Name: count, dtype: int64

In [10]:
#Check how many clients have 'pdays' that are not 999
(df['pdays'] != 999).describe()

count     41188
unique        2
top       False
freq      39673
Name: pdays, dtype: object

### EDA Findings
-**Dataset Overview:** 

41,188 rows, 21 columns

-**Target Variable (y):**

This dataset has a class imbalance favoring no: `no = 36,548`, `yes = 4,640`.

**Important Note: For model evaluation and potential class weighing**

-**Categorical Variables:**

Have a large amount of `unknowns`. Thinking about creating this as a new category or imputation.

-**Missing Values:**
No nulls. But I noticed `pdays` contains the value 999 for 39,673 out of 41,188 rows. In the dataset documentation, '-1' means the client was not previously contacted; there are no -1 values, only 999 for the same rows that would represent 'never contacted'. Next steps for this would be to create a binary indicator column and replace 999 in the pdays column with 0, so the feature can be used when modeling.

## Data Cleaning
---

In [11]:
#Handle `pdays` feature

#Create indicator for clients 'never contacted'
df['pdays_never_contacted'] = (df['pdays'] == 999).astype(int)

#Replace 999 with -1 (to follow the dataset's documentation)
df['pdays'] = df['pdays'].replace(999, -1)

#Quick check
df[['pdays', 'pdays_never_contacted']].head(10)

pdays  pdays_never_contacted
0     -1                      1
1     -1                      1
2     -1                      1
3     -1                      1
4     -1                      1
5     -1                      1
6     -1                      1
7     -1                      1
8     -1                      1
9     -1                      1

In [12]:
#Count for the number of 'unknown' (categorical features)
categorical_cols_with_unknown = ['job', 'marital', 'education', 'default', 'housing', 'loan']

for col in categorical_cols_with_unknown:
    count_unknown = (df[col] == 'unknown').sum()
    print(f"{col} - unknown count: {count_unknown}")

job - unknown count: 330
marital - unknown count: 80
education - unknown count: 1731
default - unknown count: 8597
housing - unknown count: 990
loan - unknown count: 990


### Cleaning Review
-Handled `pdays` feature: Created new indicator column titled `pdays_never_contacted` where 1 if client was never contacted (`pdays`= 999 or -1) and 0 where client was contacted at some point (`pdays ' >= 0).

-Unknown Values In Categorical Feature: Identified columns with `unknown` values and counted them. My plan is to keep `unknown` as a category for now rather than losing any info. 


## Encoding Categorical Features
---

In [13]:
#Identify the categorical columns except 'y'
cat_cols = df.select_dtypes(include='object').columns.drop('y')

#One-hot encode the categorical features
X = pd.get_dummies(df.drop(columns='y'), columns=cat_cols, drop_first=True)

#Encode 'y' as binary (Target variable)
y = df['y'].map({'no': 0, 'yes': 1})

#Quick check
print("Shape after one-hot encoding:", X.shape)
X.head(3)

Shape after one-hot encoding: (41188, 54)


age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0   56       261         1     -1         0           1.1          93.994   
1   57       149         1     -1         0           1.1          93.994   
2   37       226         1     -1         0           1.1          93.994   

   cons.conf.idx  euribor3m  nr.employed  ...  month_may  month_nov  \
0          -36.4      4.857       5191.0  ...       True      False   
1          -36.4      4.857       5191.0  ...       True      False   
2          -36.4      4.857       5191.0  ...       True      False   

   month_oct  month_sep  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0      False      False             True            False            False   
1      False      False             True            False            False   
2      False      False             True            False            False   

   day_of_week_wed  poutcome_nonexistent  poutcome_success  
0            False                  True             False  
1            False                  True             False  
2            False                  True             False  

[3 rows x 54 columns]

### Encoding Summary
-One-hot encoded: All categorical columns were converted into numeric columns. I also dropped the first category in each feature to prevent redundancy in the dataset. I noticed that after encoding, the dataset grew from 21 columns to 54 columns, and they are now all numeric. Ready for modeling.

## Data Quality Check
---

In [14]:
#Extra checks

#Verify object columns left
print(X.dtypes.value_counts())

#Check few dummy columns for 0/1 or True/False
print(X.iloc[:, -8:].head())

bool       43
int64       6
float64     5
Name: count, dtype: int64
   month_oct  month_sep  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0      False      False             True            False            False   
1      False      False             True            False            False   
2      False      False             True            False            False   
3      False      False             True            False            False   
4      False      False             True            False            False   

   day_of_week_wed  poutcome_nonexistent  poutcome_success  
0            False                  True             False  
1            False                  True             False  
2            False                  True             False  
3            False                  True             False  
4            False                  True             False  


In [15]:
#Missing Value Check
print("Total missing values:", df.isna().sum().sum())

Total missing values: 0


In [16]:
# Review all columns are numeric
print(df.dtypes.value_counts())

object     11
int64       6
float64     5
Name: count, dtype: int64


In [17]:
#Proportion (in target variable)
print(df['y'].value_counts(normalize=True))

y
no     0.887346
yes    0.112654
Name: proportion, dtype: float64


In [18]:
#Summary stats for all numeric columns
print(df.describe())

               age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593     -0.741988      0.172963   
std       10.42125    259.279249      2.770014      1.510327      0.494901   
min       17.00000      0.000000      1.000000     -1.000000      0.000000   
25%       32.00000    102.000000      1.000000     -1.000000      0.000000   
50%       38.00000    180.000000      2.000000     -1.000000      0.000000   
75%       47.00000    319.000000      3.000000     -1.000000      0.000000   
max       98.00000   4918.000000     56.000000     27.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m  \
count  41188.000000    41188.000000   41188.000000  41188.000000   
mean       0.081886       93.575664     -40.502600      3.621291   
std        1.570960        0.578840       4.628198      1.734447   
min       -3.400000      

### Data Quality Summary
-Missing Values: None

-Data Types: All categorical features are transformed to numeric (one-hot encoded)

-Target Variable: no: ~88.7% and yes: ~11.3%

**Important Note: That dataset is imblanced. Consider during modeling!**

-`pdays` feature: Handled so that -1 indicates clients 'never contacted'. `pday_never_contacted` was created (1= never contacted, 0= contacted).

-Quick check of summary stats: Campaign counts, duartion and other numeric features look to have reasonable distributions. Ages range from 17-98 with a mean of `40. 

The dataset is fully cleaned, the categorical features are encoded and the numeric features are ready for modeling. Again there is a class imbalance that will need attention. 